src/features/device-state/routes/state.ts

```diff
@@ -25,6 +26,163 @@ export const varListInsert = (varList: EnvVarList, obj: Dictionary<string>) => {
 	});
 };
 
+export type App = {
+	name: string;
+	commit: string;
+	releaseId: number;
+	releaseVersion: string;
+	appId?: string;
+	uuid?: string;
+	services: {
+		[id: number]: AppService;
+	};
+	volumes: any;
+	networks: any;
+};
+
+export type AppService = {
+	imageId: number;
+	serviceName: string;
+	image: string;
+	running: boolean;
+	environment: Dictionary<string>;
+	labels: Dictionary<string>;
+	contract?: string;
+};
+
+export type LocalState = {
+	name: string;
+	config: Dictionary<string>;
+	apps: Dictionary<Partial<App>>;
+};
+
+async function buildAppFromRelease(
+	device: AnyObject,
+	application: AnyObject,
+	release: AnyObject,
+	config: Dictionary<string>,
+): Promise<App> {
+	let composition: any = {};
+	const services: App['services'] = {};
+
+	// Parse the composition to forward values to the device
+	if (_.isObject(release.composition)) {
+		composition = release.composition;
+	} else {
+		try {
+			composition = JSON.parse(release.composition);
+		} catch (e) {
+			composition = {};
+		}
+	}
+
+	(release.contains__image as AnyObject[]).forEach((ipr) => {
+		// extract the per-image information
+		const image = ipr.image[0];
+
+		const serviceInstall = (() => {
+			const si = serviceInstallFromImage(device, image);
+
+			if (si != null) {
+				return si;
+			}
+
+			if (image.is_a_build_of__service[0] == null) {
```

Comment:
What is this check and how is it different to `si == null`?

Reply:
See my comment below.  I need some advice regarding this whole block.

Reply:
> What is this check and how is it different to si == null?

The difference is that if there were _no services installs at all_ then we would hit this error.  It is possible for `si` to be null in the case of the supervisor.

Reply:
I think the real issue is that it should not be null for the supervisor